File: articles/linear-algebra-for-systems-modeling/articles/inverse-matrices-and-structural-recovery/python/inverse_recovery.py
"""
Inverse Matrices and Structural Recovery

A small 2x2 example showing how an inverse matrix recovers a hidden state
from an observed output.
"""

def inverse_2x2(a):
    det = a[0][0] * a[1][1] - a[0][1] * a[1][0]
    if det == 0:
        raise ValueError("Matrix is singular; structural recovery is not unique.")
    return [
        [a[1][1] / det, -a[0][1] / det],
        [-a[1][0] / det, a[0][0] / det],
    ]

def matvec(a, x):
    return [
        a[0][0] * x[0] + a[0][1] * x[1],
        a[1][0] * x[0] + a[1][1] * x[1],
    ]

A = [[3, 1], [2, 4]]
x_original = [2, 1]
b = matvec(A, x_original)

A_inv = inverse_2x2(A)
x_recovered = matvec(A_inv, b)

print("A =", A)
print("Original state x =", x_original)
print("Observed output b = Ax =", b)
print("Recovered state A^-1 b =", x_recovered)

File: articles/linear-algebra-for-systems-modeling/articles/inverse-matrices-and-structural-recovery/rust/inverse_recovery.rs
fn main() {
    let a = 3.0;
    let b = 1.0;
    let c = 2.0;
    let d = 4.0;
    let y1 = 7.0;
    let y2 = 8.0;

    let det = a * d - b * c;

    if det == 0.0 {
        println!("Matrix is singular; recovery is not unique.");
        return;
    }

    let x1 = (d * y1 - b * y2) / det;
    let x2 = (-c * y1 + a * y2) / det;

    println!("Recovered state: x1 = {:.2}, x2 = {:.2}", x1, x2);
}
